### ISSUE TITLE
Harmonize Intents and Post Sockets concepts

### ISSUE THREAD

**Issue by britram** [taps-api/drafts#34]
§5.2.4, taken from Socket Intents, provides the transport application hints by binding a special higher-layer class of transport parameter to the connection. 

§7, which describes the sending API, is largely taken from Post Sockets and tends to give simpler semantics to each Content at send time.

These approaches have the similar aims, but come to them from different angles. Socket Intents was very ~Socket~Connection-oriented, and Post Sockets very ~Message~Content oriented. It seems difficult for them to coexist in a way that would allow an application developer to use this interface with ease. 

We should discuss, and agree on a way forward.

**Comment by mwelzl** [comment]
Being Content-oriented is good, as it offers more potential freedom below the api. i’m for making as much as possible content-oriented.

however: the intents things seem to be longer-term statements about app expectations or behavior - it doesn’t make much sense to combine all of them with messages.  same for some of the minset based parameters - i think having both is the only reasonable way out?!   why hard to use, when these are all optional configs on a connection handle?

**Comment by philsbln** [comment]
For Socket Intents, we already considered on which granularity these Intents can be applied – see  
https://datatracker.ietf.org/meeting/100/materials/slides-100-taps-5-draft-tiesel-taps-socketintents-01 slide 7 – I guess allowing them in any scope they are applicable is fine.

Yes, this needs to be added to the API

**Comment by philsbln** [comment]
I grabbed the ticket and will try to do a pass tomorrow – Input welcome

**Comment by britram** [comment]
I've looked over this again, and I think there are four levels of granularity here, and part of what I'm seeing as "lack of harmony" is the fact that three of them are kind of mixed up, and each of them interact with the send interface a little differently.

here's where I think we are (following my comment in #29, considering intents more deeply)...

- Intents are about what the application does. You could have them be bound to a whole context (i.e., a container for multiple connections; the process level object), except some applications might need to speak multiple protocols with wildly different profiles (e.g., webrtc), so it's probably a useful simplifying decision to bind them only to Connections. Since Intents are advisory only, violating an Intent while sending should never be a runtime error.

- Transport selection properties are basically only about the constraints on Protocol Stack selection, and become irrelevant once the stack is selected (we can have different discussion about how to do re-pre-establishment and whether or not the perceived need to do that is an application/API design error up in #38). Certain violations of transport selection properties may be runtime errors, but the API should endeavor to gracefully degrade. 

- Transport stack properties are a lower-level way to reach down into the stack and tweak performance parameters, though these should be expressed in a transport-independent way when possible. Inconsistent transport stack property configuration will lead to runtime errors, but this is the application's fault for being picky. (I suspect servers will need to do more of this sort of thing than clients).

- Send parameters are about what should be done with this particular piece of Content within the context of the Connection it's sent on. I think we've got the right set of these, though we still need to come to consensus on priorities, it seems...

**Comment by philsbln** [comment]
I disagree with this perspective because it drops path selection from consideration.

- Intents are equally useful for path setup as well as assigning contents to paths – Theresa and I deeply looked into this.\
Therefore we should be able to express intents when sending contents. If the intent expressed on a content contradicts the path set up, the transport system must ignore them and continue.
- It might also make sense to change path selection parameters (see Issue #38)

In addition to that, there are Transport Stack Properties that apply on a per-conetnts granularity  (e.g. "urgent data" in TCP) - but these are 

I don't like the idea to add the whole "require/prefer/avoid/prohibit" for all Send Parameters – Intents and all examples of Transport Stack Properties I came up with make no use of them.
What we can to harmonize the concepts instead changing the the Protocol- and Path Selection Properties into Key/Values and put the "require/prefer/avoid/prohibit" into the value field

~~~
transportParameters.addProperty(ReliableDataTransfer, require)
~~~

**Comment by britram** [comment]
> I disagree with this perspective because it drops path selection from consideration.

I don't see how, can you say more here?

> Intents are equally useful for path setup as well as assigning contents to paths

So perhaps the issue here is that we're artificially trying to keep Intents separate from other path and protocol selection parameters, and we should try to recast all of these in terms of "selection parameters" attached to a preference.

> I don't like the idea to add the whole "require/prefer/avoid/prohibit" for all Send Parameters

Ugh. Me either, but I don't think I suggested that.

> What we can to harmonize the concepts instead changing the the Protocol- and Path Selection Properties into Key/Values and put the "require/prefer/avoid/prohibit" into the value field

hm, this is moving in the right direction here, but the interface still feels a little C89ish. How about making the preference the action on the parameter object instead? i.e.:

```
transportParameters.Require(AlwaysReliableDataTransfer)
transportParameters.Require(OrderedDataTransfer)
transportParameters.Prefer(Establish0RTT)
transportParameters.Intend(ConstantDataRate, 300kbps)
```

and so on?

**Comment by philsbln** [comment]
>> I disagree with this perspective because it drops path selection from consideration.
>
>I don't see how, can you say more here?

I disagree with excluding Intents (and some of the Path Selection Properties) from the Contents Send Properties.

> > Intents are equally useful for path setup as well as assigning contents to paths
>
> So perhaps the issue here is that we're artificially trying to keep Intents separate from other path and protocol selection parameters, and we should try to recast all of these in terms of "selection parameters" attached to a preference.

Agreed – but I was arguing against artificially trying to keep Selection Parameters and Send Parameters apart. There is an intersection between both sets and that intersection should neither be duplicated nor look awkward in the API.

~~~
transportParameters.Require(AlwaysReliableDataTransfer)
transportParameters.Require(OrderedDataTransfer)
transportParameters.Prefer(Establish0RTT)
transportParameters.Intend(ConstantDataRate, 300kbps)
~~~

I like this. How do we design the ContentsParameters then? 
I'd propose the following:

~~~
sentParameters := connection.getDefaultSendParameters()
sentParameters.Prefer(Sent0RTT)
sentParameters.Intend(RecvContentSize, 2MB)
connection.Send(contentRef, sentParameters)
~~~

**Comment by britram** [comment]
> I disagree with excluding Intents (and some of the Path Selection Properties) from the Contents Send Properties.

Ah, okay. Hm. Yes, we disagree. :)

The send parameters *really* should be limited to those things which are bound to the Content. Indeed, one way to do this would for *content* to have Parameters bound to it, but that leads IMO to an unpleasant-to-use interface. In some languages, there's no way around that, but as @mwelzl points out some languages (e.g. Python) provide optional arguments, and in those cases send parameters can be optional args to the `send()` function. We shouldn't define something that mandates a clunkier interface than the idiom of the language allows.

We can of course define send parameters that are related to transport parameters (and indeed, we should provide some easy to use reference to see which of these parameters are related / exclusive / etc), but I think they're fundamentally different animals.

**Comment by theri** [comment]
I'm a bit confused here. So far I assumed that Intents could be set both per Connection and on individual Content, so they could be both transportParameters and sendParameters. For me, transportParameters and sendParameters are different things, but they can still be expressed using the same vocabulary, as discussed in #29.
So the disagreement here is on whether an application should also be allowed set Intents as sendParameters, right?

**Comment by britram** [comment]
> So the disagreement here is on whether an application should also be allowed set Intents as sendParameters, right?

I think so, yes. In document terms, the two ways forward seem to be:

- Keep "Intents" as a separate category of things in §5.2.4, and note in §7 that all Intents are also usable as send parameters (somehow).

- Eliminate §5.2.4 as a separate entity, fold intents into transport parameters, and create intents-derived send parameters where appropriate.

I'm proposing the latter.

There's another discussion we need to have about which Intents stay in the API and which don't -- some of these, for instance, seem difficult to non-arbitrarily map down to behaviors in the underlying protocol stacks as @mwelzl pointed out in mail -- but that's a separate issue.

**Comment by theri** [comment]
I see, thanks for clarifying.
In the current §5.2.4, it says that some Intents can also apply to individual messages, namely Size to be Sent / Received, Timeliness, Disruption Resilience, and Cost Preferences. This reflects the previous work on granularity that Philipp pointed out in his earlier post in this thread.
Creating Intents-derived send parameters for (a subset of) the above sounds feasible to me. 
About folding the Connection-based Intents into the other transport parameters, I see two potential issues which I'm not sure yet how to resolve:
- Under what heading to put them - As they can influence both protocol and path selection, neither of these heading seemed appropriate for me, that's why I put them under a separate heading. If not Intents, what should this heading be?
- Reformulating them - Philipp and I already discussed whether some Intents could be reformulated as properties that fit into the "Require", "Prefer", "Avoid", "Prohibit" preferences. But not all Intents can fit this scheme. What to do with those?

**Comment by britram** [comment]
One quick point:

> Philipp and I already discussed whether some Intents could be reformulated as properties that fit into the "Require", "Prefer", "Avoid", "Prohibit" preferences

Though the phrasing sounds silly, "Intend" semantics are pretty much the same as "Prefer" semantics. ("I intend to use this connection in this way" -> "I prefer PSIs/paths that work well with this assumption")

**Comment by tfpauly** [comment]
I agree with Brian that the properties in the Application Intents section should be distributed between the other properties sections. If we want a new "Data/Content Properties" that could be good too, for the things that are more about what the application wants for its data.

I'd say for you list:

- Traffic Category or Service Class seems like a Protocol Selection Property (previously Generic Protocol Property). This is a generic protocol option that influences the nature of the stream or flow I have. It influences how I initialize my transport protocol.

- Size/Duration/Timeliness/Resilience all seem like Protocol Selection Properties (previously Generic Protocol Properties). Again, these will help choose which protocols to use, and how to prep and schedule them.

- Cost Preferences (for us, Allow/Prohibit Expensive) is a Path Selection Property.

One thing that would be good to note is that Path Selection properties take effect not just in the initial bringup, but for all of the evaluation of flows and interfaces for multipath or migratable protocols.

I would argue again that any properties per-send/receive should be specified as separate APIs from the pre-establishment APIs, even if they look similar, since they have different effects and domains of relevance.

**Comment by theri** [comment]
For Traffic Category and Cost Preferences, I can see them fit better into Protocol or Path Selection, but I'm not sure yet if they fit exclusively. For the other Intents I'm not convinced that they can be sorted at all. For example, so far we use Size mainly to select the path on which we predict the shortest completion time for the transfer, i.e., depending on whether it is a "large" or a "small" transfer. For Timeliness and Resilience I can also more easily think of a path selection example based on lower latency and availability, respectively. On the other hand, the same Intents can help choose and configure transport protocols. So maybe putting them as "Data/Content Properties" would be better than trying to squeeze them into one of the two Selection categories.
I think making the above clear in the Implementation draft will help us in this discussion and I'm planning to get started on this soon.
I agree with Tommy's point about Path Selection properties also taking effect later on.
About the per-send/receive properties, Philipp is currently working on Intents-derived per-Content properties, see #38.

**Comment by mwelzl** [comment]
I agree that the intents fit in different parts. I think that's part of the problem - they're treated (by being under one subheading) as a category in their own right, but they're really quite a heterogeneous set of things.

**Comment by philsbln** [comment]
I think the whole dividing into path selection properties, protocol selection properties, intents and protocol properties is rather historic, than making sense at all.
In my world, we have three Dimensions of ~~properties~~ parameters:

1. Data Type
   - Int/Enum
      - Protocol Properties
      - Intents
   - Preference Enum
      - Preferences
2. Strength of Expression
   - advisory
      - Intents 
      - Preferences with avoid/prefer preference 
      - Protocol Properties
      - Content Properties
   - mandatory
      - Preferences with require/prohibit preference (advisory, somehow boolean)
      - Protocol Properties
      - Content Properties
3. Effected Subject
   - Protocol Selection
      - Preferences
      - Intents
   - Protocol Parameters
      - Preferences
      - Protocol Properties
      - Intents
   - Path Selection (read MPTCP subflow)
      - Preferences
      - Protocol Properties
      - Intents
   - Path Selection (read MPTCP segment to subflow assignment)
      - Preferences
      - Protocol Properties
      - Content Properties
      - Intents
   - Content Transmission
      - Preferences
      - Protocol Properties
      - Intents
4. Time of use
   - Pre-Establishment
      - Preferences
      - Protocol Properties
      - Intents
   - Establishment
      - Preferences
      - Protocol Properties
      - Intents
   - Content Send
      - Preferences
      - Content Properties
      - Protocol Properties
      - Intents

Mixing these dimensions in terminology is bound to fail (as we currently try to do).
Still working out how and where to fix this

**Comment by philsbln** [comment]
First consequences:
- get rid of differentiating between Path Selection and Protocol Selection properties
- rename Path Selection and Protocol Selection properties to Preferences
- remove the Preference parameter from Protocol properties and Intents
- Make calls for Transport and Send Properties roughly the same

**Comment by britram** [comment]
> remove the Preference parameter from Protocol properties and Intents

wait, why? Whether or not a parameter is preferred (or avoided, or required, or prohibited) is orthogonal to how we talk about them, or am I missing something here?

**Comment by philsbln** [comment]
Look at #protocol-props – there is nothin to require/prefer/avoid/prohibt

**Comment by britram** [comment]
ahhh okay, got it.

**Comment by britram** [comment]
I had a brief look at the branch here, and I'm very confused about what you're trying to do here.

7.1.4.1. seems unnecessary as a send parameter; the length of the Content gives the size to be sent. *Maybe* you could tell the transport the size of the message it should expect to receive based on a given send, but I'm not sure I know what a transport will do with it

7.1.4.2. is similarly odd; certainly this is a property of multiple sends, not a single one?

7.1.4.3. seems to be an alternate expression of Lifetime.

7.1.4.4. seems to be an alternate expression of path selection property.

I think we should probably take a step back here, and instead of starting from an attempt to map Socket Intents concepts directly into this document, attempt to see what concepts in Socket Intents are missing from the API, and how those can best be added in keeping with the architecture (and can be implemented using transport services described in 8095 and the upcoming security properties doc). In any case I don't think this will converge before -00, so let's take the time pressure off and mark it post-London...

**Comment by philsbln** [comment]
This was "just" a copy an paste and is fixed now.

Most work went into re-structuring and fixing terminology to harmonise Sent and Trasnport Parameter concepts.
I will file a PR today and remove the post-London milestone for from this ticket. 

I am fine with a partial merge and a new post-London issues for discussing Send Parameter Intents.

**Comment by britram** [comment]
this is closeable with the merge of #48, yes?

**Comment by philsbln** [comment]
Most things discussed here are addressed with merge #48 - remaining things are post-London and covered by issue #38
